### PR TITLE
612 Introduced Percolator Recap Search Alerts 

### PR DESCRIPTION
**Pull request by albertisfu**

### CODE REVIEW

**Comment by albertisfu** [comment]
When adding grouping for OA Search Alerts, I noticed that I missed omitting the sending of RT email alerts for non-members for RECAP Alerts.

I've fixed this issue and tweaked the tests.

Note that webhooks in RECAP as in OA are always triggered regardless of the user's membership status.

Additionally, I noticed an error in ES (`document_parsing_exception`) when indexing queries using advanced syntax with fields that are not contained in the auxiliary indices. This error is not being caught by tests. So I will analyze if this is an issue in terms of filtering queries and, if so, look for an alternative approach.

**Comment by albertisfu** [comment]
The issue related to the `document_parsing_exception` error is now solved.

The reason it was not being caught by tests was because the test query that was triggering the error:
`q="405 Civil" AND pacer_doc_id:018036652436&type=r`

Fell into the following case in the auxiliary percolator queries boolean table:

Alert matched in DocketDocumentPercolator | Alert matched in RECAPDocumentPercolator | Trigger Alert | Description
-- | -- | -- | --
False | False | True | AND Cross-object queries

As a result, none of the auxiliary queries required returning results.

The test that could have failed was:
`q="plain text for 018036652436" OR cause:"405 Civil"`

But it didn't fail because the alert indexing issue was only related to advanced queries using a field and its value without quotes `pacer_doc_id:018036652436` if the field value contained quotes like cause:"405 Civil" it was accepted by the ES parser.

So I updated that test to `q="018036652436" OR cause:405&type=r`.

And I could confirm that this issue indexing some edge cases alerts could impact Alerts filtering.

### The solution:

After analyzing the problem, I determined that the best approach to solve this issue was to change the strategy regarding auxiliary percolator queries. Instead of creating two additional Percolator indices to save RECAPDocument-only queries and Docket-only queries and percolating the full document, it's better to use a single Percolator index. 
This index is the same for the main document where full alerts are stored. For the auxiliary percolator queries, instead of percolating the full document, we can percolate one document with only child fields and another percolator query with only parent fields.
This way, auxiliary queries can tell us when a document is a Docket-only query or not, allowing us to filter out RECAPDocuments.

This approach also simplifies the code, as we no longer need the additional `RECAPDocumentPercolator` and `DocketDocumentPercolator` indices, so they have been removed.

File: cl/alerts/migrations/0010_add_schedule_alert_hit_content_type.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            name="object_id",
+            field=models.PositiveIntegerField(null=True),
+        ),
+        migrations.AddIndex(
+            model_name="scheduledalerthit",
+            index=models.Index(
+                fields=["content_type", "object_id"],
+                name="alerts_sche_content_c5e627_idx",
+            ),
+        ),
```
Comment (ERosendo) [inline cl/alerts/migrations/0010_add_schedule_alert_hit_content_type.py:36]: Since both fields are new, the index creation process should be relatively quick. however, to make sure we don't slow down anything else while it's working, let's use `AddIndexConcurrently`.

**Review by ERosendo — CHANGES_REQUESTED** [review]
@albertisfu, here are some suggestions for minor improvements to this pull request.

File: cl/alerts/migrations/0010_add_schedule_alert_hit_content_type.py (hunk 1 above)
Comment (albertisfu) [reply cl/alerts/migrations/0010_add_schedule_alert_hit_content_type.py]: Done! I split the migrations into two files: one for adding the new fields and another for the concurrent index, since it needs to be outside of a transaction (atomic = False)

File: cl/search/types.py [hunk 2]
```diff
@@ -97,10 +99,17 @@
 
 ESDictDocument = dict[str, Any]
 
+# TODO: Remove after scheduled OA alerts have been processed.
 PercolatorResponseType = tuple[list[Hit], ESDictDocument]
 
+PercolatorResponsesType = tuple[
+    list[Hit], list[Hit], list[Hit], ESDictDocument, str
+]
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/search/types.py]: Sure, refactored as a dataclass

File: cl/alerts/tasks.py [hunk 3]
```diff
@@ -626,6 +642,180 @@ def process_percolator_response(response: PercolatorResponseType) -> None:
         logger.info(f"Sent {alerts_sent} {Alert.REAL_TIME} email alerts.")
 
 
+@app.task(ignore_result=True)
+def percolator_response_processing(response: PercolatorResponsesType) -> None:
+    """Process the response from the percolator and handle alerts triggered by
+     the percolator query.
+
+    :param response: A two tuple, a list of Alerts triggered and the document
+    data that triggered the alert.
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/alerts/tasks.py:651]: Yeah I updated the docstring.

File: cl/alerts/utils.py [hunk 4]
```diff
@@ -186,3 +504,217 @@ def has_document_alert_hit_been_triggered(
...
+            r_fields_to_ignore = {"absolute_url"}
+            child_document_content = select_es_document_fields(
+                ESRECAPBaseDocument, document_content_plain, r_fields_to_ignore
+            )
+
+            d_fields_to_ignore = {
+                "docket_slug",
+                "docket_absolute_url",
+                "court_exact",
+                "docket_child",
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/alerts/utils.py]: Removed

File: cl/alerts/utils.py [hunk 5]
```diff
@@ -186,3 +504,217 @@ def has_document_alert_hit_been_triggered(
...
+    alert_in_docket_hits = alert_id in docket_hits
+    if not alert_in_docket_hits and not alert_in_rd_hits:
+        return True
+    elif not alert_in_docket_hits and alert_in_rd_hits:
+        return True
+    elif alert_in_docket_hits and not alert_in_rd_hits:
+        return False
+    elif alert_in_docket_hits and alert_in_rd_hits:
+        return True
+    return False
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/alerts/utils.py]: Changed

File: cl/alerts/utils.py [hunk 6]
```diff
@@ -100,6 +144,213 @@ def percolate_document(
...
+    percolator_index: str,
+    document_index: str | None = None,
+    documents_to_percolate: (
+        tuple[ESDictDocument, ESDictDocument, ESDictDocument] | None
+    ) = None,
+    app_label: str | None = None,
+    main_search_after: int | None = None,
+    rd_search_after: int | None = None,
+    d_search_after: int | None = None,
+) -> tuple[Response, Response | None, Response | None]:
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/alerts/utils.py]: Yeah added `PercolatorResponses` dataclass

File: cl/search/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/search/types.py]: This is called now SendAlertsResponse

File: cl/alerts/utils.py [hunk 7]
```diff
@@ -100,6 +144,213 @@ def percolate_document(
...
+    main_alerts_returned = len(main_response.hits.hits)
+    rd_response = d_response = None
+    if initial_responses[1]:
+        rd_response = initial_responses[1]
+        all_rd_alert_hits.extend(rd_response.hits)
+    if initial_responses[2]:
+        d_response = initial_responses[2]
+        all_d_alert_hits.extend(d_response.hits)
+
+    if main_total_hits > settings.ELASTICSEARCH_PAGINATION_BATCH_SIZE:
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/alerts/utils.py]: changed!

File: cl/lib/es_signal_processor.py [hunk 8]
```diff
@@ -191,17 +191,20 @@ def update_es_documents(
...
+                            ),
+                            (compose_app_label(instance), instance.pk),
+                            fields_map,
                         ),
-                        (compose_app_label(instance), instance.pk),
-                        fields_map,
-                    )
+                        send_or_schedule_search_alerts.s(),
+                        percolator_response_processing.s(),
+                    ).apply_async()
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/lib/es_signal_processor.py]: Yeah, that's right! Thanks for catching this potential issue. I've switched from using lambda to partial instead.

**Comment by albertisfu** [comment]
@ERosendo thank you for your comments and suggestions, this is ready for another review!

**Comment by albertisfu** [comment]
I made an additional change. We had `PERCOLATOR_SEARCH_ALERTS_ENABLED` to disable the percolator via a setting that was `False` by default. However, this setting was also disabling Oral Arguments Search Alerts, which are currently enabled in production.
Therefore, I renamed the setting to `PERCOLATOR_RECAP_SEARCH_ALERTS_ENABLED`, so it now only disables percolation related to RECAP Alerts.

@mlissner when this is merged the percolator will be disabled for RECAP Alerts. then it's required to create the `RECAPPercolator` index  for doing that:

```
from cl.search.documents import RECAPPercolator
RECAPPercolator.init()
```

If, for any reason, you need to delete the Percolator index, you can run:

`RECAPPercolator._index.delete(ignore=404)`

Additionally, to enable the sending of RT alerts every 5 minutes, it's required to set up the daemon `cl-send-rt-percolator-alerts`, which is already configured in `docker-entrypoint.sh`

Once that's in place, you can enable the `PERCOLATOR_RECAP_SEARCH_ALERTS_ENABLED` setting to start the percolation of documents related to RECAP.

**Review by ERosendo — APPROVED** [review]
LGTM
